src/Microsoft.OData.Client/Materialization/FeedAndEntryMaterializerAdapter.cs

```diff
@@ -296,6 +323,83 @@ private ODataResourceSet ReadFeedCore(bool lazy)
             return result;
         }
 
+        /// <summary>
+        /// Reads an <see cref="ODataDeltaResourceSet"/>.
+        /// </summary>
+        /// <returns>The <see cref="MaterializerDeltaFeed"/> of the read <see cref="ODataDeltaResourceSet"/>.</returns>
+        private MaterializerDeltaFeed ReadDeltaFeedCore()
+        {
+            this.ExpectState(ODataReaderState.DeltaResourceSetStart);
+
+            ODataDeltaResourceSet result = (ODataDeltaResourceSet)this.reader.Item;
+            MaterializerDeltaFeed feed = null;
+
+            if (result == null)
+            {
+                return null;
+            }
+
+            feed = MaterializerDeltaFeed.CreateDeltaFeed(
+                    result,
+                    new List<IMaterializerState>(),
+                    this.materializerContext);
+
+            if (this.items?.Count > 0)
+            {
+                this.items.Push(feed);
+            }
+            else
+            {
+                IMaterializerState topLevelItem = feed;
+                this.items?.Push(topLevelItem);
+            }
+
+            do
+            {
+                this.AssertRead();
+
+                switch (this.reader.State)
+                {
+                    case ODataReaderState.ResourceStart:
+                        MaterializerEntry entry = this.ReadEntryCore();
+                        feed.AddEntry(entry);
+                        break;
+                    case ODataReaderState.DeletedResourceStart:
+                        MaterializerDeletedEntry deletedMaterializerEntry = this.ReadDeletedResource();
+                        feed.AddEntry(deletedMaterializerEntry);
+                        break;
+                    case ODataReaderState.DeletedResourceEnd:
+                    case ODataReaderState.ResourceEnd:
+                    case ODataReaderState.DeltaResourceSetEnd:
+                        this.items?.Pop();
+                        break;
+                    default:
+                        throw DSClient.Error.InternalError(InternalError.UnexpectedReadState);
+                }
+            }
+            while (this.reader.State != ODataReaderState.DeltaResourceSetEnd);
+
+            return feed;
+        }
+
+        /// <summary>
+        /// Adds an <see cref="IMaterializerState"/> entry to a parent entry.
+        /// </summary>
+        /// <param name="materializerState">The <see cref="IMaterializerState"/> entry being added to the parent item.</param>
+        private void AddResourceToParent(IMaterializerState materializerState)
+        {
+            IMaterializerState topItem = this.items?.Peek();

```

Comment:
I think you should add `Debug.Assert(this.items != null, "yada yada");` to this method and eliminate the null propagation operator. Otherwise if it's null and no item is pushed onto the stack, would the rest of the logic in this method be relevant/valid?

Reply:
the stack is only used when we are reading a delta payload. If we are just reading a single payload then this.items will be null hence the check.